src/block_cache/block_chunkcache.c

```diff
@@ -0,0 +1,552 @@
+ */
+int
+__wt_chunkcache_get(WT_SESSION_IMPL *session, WT_BLOCK *block, uint32_t objectid, wt_off_t offset,
+  uint32_t size, void *dst)
+{
+    WT_CHUNKCACHE *chunkcache;
+    WT_CHUNKCACHE_CHUNK *chunk;
+    WT_CHUNKCACHE_HASHID hash_id;
+    WT_DECL_RET;
+    bool chunk_cached;
+    unsigned int bucket_id, retries;
+    size_t already_read, left_to_read, readable_in_chunk, size_copied;
+
+    chunkcache = &S2C(session)->chunkcache;
+    already_read = 0;
+    left_to_read = size;
+    retries = 0;
+
+    if (!chunkcache->configured)
+        return (ENOTSUP);
+
+    __wt_verbose(session, WT_VERB_CHUNKCACHE, "get: %s(%u), offset=%" PRId64 ", size=%u",
+      (char *)block->name, objectid, offset, size);
+
+    WT_STAT_CONN_INCR(session, chunk_cache_lookups);
+
+    /* A block may span two (or more) chunks. Loop until we have read all the data. */
+    while (left_to_read > 0) {
+        /* Find the bucket for the chunk containing this offset. */
+        bucket_id = __chunkcache_make_hash(
+          chunkcache, &hash_id, block, objectid, offset + (wt_off_t)already_read);
+retry:
+        chunk_cached = false;
+        __wt_spin_lock(session, WT_BUCKET_LOCK(chunkcache, bucket_id));
+        TAILQ_FOREACH (chunk, WT_BUCKET_CHUNKS(chunkcache, bucket_id), next_chunk) {
+            if (memcmp(&chunk->hash_id, &hash_id, sizeof(hash_id)) == 0) {
+                /* If the chunk is there, but invalid, there is I/O in progress. Retry. */
+                if (!chunk->valid) {
+                    __wt_spin_unlock(session, WT_BUCKET_LOCK(chunkcache, bucket_id));
+                    if (retries++ < WT_CHUNKCACHE_MAX_RETRIES) {
+                        WT_STAT_CONN_INCR(session, chunk_cache_retries);
+                        __wt_yield();
+                        goto retry;
+                    } else {
+                        __wt_verbose(session, WT_VERB_CHUNKCACHE,
+                          "lookup timed out after %u retries", retries);
+                        WT_STAT_CONN_INCR(session, chunk_cache_toomany_retries);
+                        return (EAGAIN);
+                    }
+                }
+                /* Found the needed chunk. */
+                chunk_cached = true;
+                WT_ASSERT(session,
+                  WT_BLOCK_OVERLAPS_CHUNK(chunk->chunk_offset, offset + (wt_off_t)already_read,
+                    chunk->chunk_size, left_to_read));
+
+                /* We can't read beyond the chunk's boundary. */
+                readable_in_chunk =
+                  (size_t)chunk->chunk_offset + chunk->chunk_size - (size_t)offset;
+                size_copied = WT_MIN(readable_in_chunk, left_to_read);
+                memcpy((void *)((uint64_t)dst + already_read),
+                  chunk->chunk_memory + (offset + (wt_off_t)already_read - chunk->chunk_offset),
+                  size_copied);
+                __wt_spin_unlock(session, WT_BUCKET_LOCK(chunkcache, bucket_id));
```

Comment:
Do we want to put this chunk at the top of the LRU list (i.e. Most recently used)?  (Might make sense in the context of my comment in __wt_chunkcache_remove).

Reply:
Oh totally! I had that logic before, but then it got lost as I was refactoring. Thanks for catching this.